ledger-service/http-json/src/main/scala/com/digitalasset/http/domain.scala

```diff
@@ -131,6 +133,29 @@ object domain extends com.daml.fetchcontracts.domain.Aliases {
 
   final case class PartyDetails(identifier: Party, displayName: Option[String], isLocal: Boolean)
 
+  final case class UserRights(canActAs: List[Party], canReadAs: List[Party], isAdmin: Boolean)
+
+  object UserRights {
+    def fromListUserRights(input: Vector[UserRight]): UserRights = {
+      val (canActAs, remaining1) = input.partitionMap {
+        case CanActAs(party) => Left(party)
+        case other => Right(other)
+      }
+      val (canReadAs, remaining2) = remaining1.partitionMap {
+        case CanReadAs(party) => Left(party)
+        case other => Right(other)
+      }
+      val isAdmin = remaining2.exists {
+        case ParticipantAdmin =>
+          true
+        case _ => false
```

Comment:
```suggestion
  
```

Can skip the wildcard check on exists

Reply:
This isn't allowed.

```scala
scala> Vector(1, 2) exists {case 2 => true}
scala.MatchError: 1 (of class java.lang.Integer)
```

Reply:
There's an interesting broader semantic point here, which plays into the compiler warning telling you about this problem (albeit indirectly):

1. If the _expected type_ of the lambda is `A => B` for some _A_, _B_, exhaustiveness checking applies to a `case` style lambda.
2. If the _expected type_ of the lambda is `A PartialFunction B`, exhaustiveness checking _does not_ apply (i.e. missing cases are fine).

This is _counterintuitive_ because for all _X_, _Y_, `(X PartialFunction Y) <: (X => Y)`, and therefore not permitting an expression for the latter that is permitted for the former seems to violate the Liskov Substitution Principle.

In practice, however, a function expecting the (2) form is prepared to deal with missing cases in the lambda, whereas one expecting the (1) form is _not_ prepared so, so this warning behavior is almost always what you really want. For example, you get the warning because `exists` expects a `=>`, not a `PartialFunction`, and indeed as you see, is not prepared to deal with missing `case`s.

Reply:
TIL, thanks for the explanation 👍